src/main/java/com/mongodb/kafka/connect/source/heartbeat/HeartbeatManager.java

```diff
@@ -61,36 +61,43 @@ public HeartbeatManager(
 
   public Optional<SourceRecord> heartbeat() {
     if (cursor == null) {
+      LOGGER.debug("Returning no heartbeat: null cursor");
       return Optional.empty();
     }
-
     long currentMS = time.milliseconds();
     long timeSinceHeartbeatMS = currentMS - lastHeartbeatMS;
-
-    if (canCreateHeartbeat && timeSinceHeartbeatMS > heartbeatIntervalMS) {
-      lastHeartbeatMS = currentMS;
-      return Optional.ofNullable(cursor.getResumeToken())
-          .map(
-              r -> {
-                String resumeToken = r.toJson();
-                if (!resumeToken.equals(lastResumeToken)) {
-                  LOGGER.info("Generating heartbeat event. {}", resumeToken);
-                  Map<String, String> sourceOffset = new HashMap<>();
-                  sourceOffset.put(ID_FIELD, resumeToken);
-                  sourceOffset.put(HEARTBEAT_KEY, "true");
-                  lastResumeToken = resumeToken;
-                  return new SourceRecord(
-                      partitionMap,
-                      sourceOffset,
-                      heartbeatTopicName,
-                      Schema.STRING_SCHEMA,
-                      resumeToken,
-                      Schema.OPTIONAL_BYTES_SCHEMA,
-                      null);
-                }
-                return null;
-              });
+    if (!canCreateHeartbeat) {
+      LOGGER.debug("Returning no heartbeat: canCreateHeartbeat is false");
+      return Optional.empty();
+    }
+    if (timeSinceHeartbeatMS <= heartbeatIntervalMS) {
+      LOGGER.debug("Returning no heartbeat: timeSinceHeartbeat has not exceeded heartbeatInterval");
+      return Optional.empty();
+    }
+    lastHeartbeatMS = currentMS;
+    BsonDocument resumeTokenBson = cursor.getResumeToken();
+    if (resumeTokenBson == null) {
+      LOGGER.debug("Returning no heartbeat: cursor resumeToken is null");
+      return Optional.empty();
+    }
+    String resumeToken = resumeTokenBson.toJson();
+    if (resumeToken.equals(lastResumeToken)) {
+      LOGGER.debug("Returning no heartbeat: same resumeToken");
+      return Optional.empty();
     }
-    return Optional.empty();
+    LOGGER.info("Generating heartbeat event. {}", resumeToken);
```

Comment:
What's the rationale for info level here?

Reply:
The message is preserved from the original happy-path code at line 77. I left it alone, though I can see how it might be argued to be either info or debug.